articles/contributing/style-guide.md

```diff
@@ -0,0 +1,451 @@
+- When a function or operation is used only in when partially applied, ensure that its name ends with `Impl`.
+  Alternatively, refactor to make the `Impl` callable useful in its own right.
+
+# [Examples](#tab/examples)
+
+|   | Name | Description |
+|---|------|-------------|
+| ☒ | <s>`ApplyDecomposedOperation_`</s> | The underscore `_` should not appear at the end of the name. |
+| ☑ | `_ApplyDecomposedOperation` | The underscore `_` at the beginning clearly indicates that this operation is for internal use only. |
+| ☑ | `ComposeImpl` | The suffix `Impl` clearly indicates that this callable is meant to support the `Compose` callable. |
+
+***
+
+### Variants ###
+
+Though this limitation may not persist in future versions of Q#, it is presently the case that there will often be groups of related operations or functions that are distinguished by which functors their inputs support, or by the concrete types of their arguments.
+These groups can be distinguished by using the same root name, followed by one or two letters that indicate its variant.
+
+| Suffix | Meaning |
+|--------|---------|
+| `A` | Input expected to support `Adjoint` |
+| `C` | Input expected to support `Controlled` |
+| `CA` | Input expected to support `Controlled` and `Adjoint` |
+| `I` | Input or inputs are of type `Int` |
+| `D` | Input or inputs are of type `Double` |
+| `L` | Input or inputs are of type `BigInt` |
+| `BE` | Input or inputs are of type `BigEndian` |
+| `LE` | Input or inputs are of type `LittleEndian` |
+
+# [Guidance](#tab/guidance)
+
+We suggest:
+
+- If a function or operation is not related to any similar functions or operations by the types and functor support of their inputs, do not use a suffix.
+- If a function or operation is related to any similar functions or operations by the types and functor support of their inputs, use suffixes as in the table above to distinguish variants.
+
+# [Examples](#tab/examples)
+
+***
+
+### Arguments and Variables ###
+
+A key goal of the Q# code for a function or operation is for it to be easily read and understood.
+Similarly, the names of inputs and type arguments should communicate how a function or argument will be used once provided.
+
+
+# [Guidance](#tab/guidance) 
+
+We suggest:
+
+- For all variable and input names, use `pascalCase` in strong preference to `CamelCase`, `snake_case`, or `ANGRY_CASE`.
+- Input names should be descriptive; avoid one or two letter names where possible.
+- Operations and functions accepting exactly one type argument should denote that type argument by `T` when its role is obvious.
+- If a function or operation takes multiple type arguments, or if the role of a single type argument is not obvious, consider using a short capitalized word prefaced by `T` (e.g.: `TOutput`) for each type.
+- Do not include type names in argument and variable names unless absolutely necessary; this information can and should be provided by your development environment.
```

Comment:
I am lacking imagination here regarding a scenario that qualified as "absolutely necessary"?

Reply:
I guess I meant for that as a hedge for my own lack of imagination, but I think that's already covered by the initial rule so I'll drop it here.